File: 0044.Wildcard_Matching.cpp
//94.54%RT

class Solution {
public:
    bool isMatch(string s, string p) {
        return isMatch(s.c_str(), p.c_str());
    }

private:
    bool isMatch(const char *s, const char *p) {
        const char* star = nullptr; // Track the position of the last '*' encountered
        const char* ss = s; // Track the position of the string before the last '*'

        while (*s) { // Loop until the end of the string
            // Advance both pointers if the characters match or '?' is found in the pattern
            if ((*p == '?') || (*p == *s)) {
                s++;
                p++;
                continue;
            }

            // If '*' is found in the pattern, track its position and advance the pattern pointer
            if (*p == '*') {
                star = p++;
                ss = s;
                continue;
            }

            // If the characters don't match, but a '*' was previously encountered, backtrack to the '*'
            if (star) {
                p = star + 1;
                s = ++ss;
                continue;
            }

            // If the characters don't match and no '*' was previously encountered, it's a mismatch
            return false;
        }

        // Check for remaining '*' characters in the pattern
        while (*p == '*') {
            p++;
        }

        // Return true if all characters in the pattern have been matched
        return !*p;
    }
};

File: 0001-0100/0044.Wildcard_Matching.java
//95.27%RT

class Solution {
    public boolean isMatch(String s, String p) {
        return comparison(s, p);
    }

    boolean comparison(String str, String pattern) {
        int s = 0, p = 0, match = 0, starIdx = -1;

        // Loop through the string and pattern
        while (s < str.length()) {
            // If the current characters match or the pattern character is '?'
            if (p < pattern.length() && (pattern.charAt(p) == '?' || str.charAt(s) == pattern.charAt(p))) {
                s++;
                p++;
            }
            // If the current pattern character is '*'
            else if (p < pattern.length() && pattern.charAt(p) == '*') {
                // Save the position of the '*' and the current string position
                starIdx = p;
                match = s;
                p++;
            }
            // If the last pattern character was '*'
            else if (starIdx != -1) {
                // Advance the pattern pointer to after the '*'
                p = starIdx + 1;
                // Advance the string pointer to after the matched characters
                match++;
                s = match;
            }
            // If the current pattern character is not '*' and the last pattern character was not '*',
            // and the characters do not match
            else {
                return false;
            }
        }

        // Check for remaining '*' characters in the pattern
        while (p < pattern.length() && pattern.charAt(p) == '*') {
            p++;
        }

        // Return true if all characters in the pattern have been matched
        return p == pattern.length();
    }
}
